tests/qtrajectory_test.cc

```diff
@@ -0,0 +1,439 @@
+  using Ry = Cirq::ry<fp_type>;
+
+  auto normal = KrausOperator<types::Gate>::kNormal;
+
+  ncircuit.push_back({{normal, 1, 1.0, {Hd::Create(0, 0)}}});
+  ncircuit.push_back({{normal, 1, 1.0, {Hd::Create(0, 1)}}});
+  add_noise(1, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {IS::Create(2, 0, 1)}}});
+  add_noise(3, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {Rx::Create(4, 0, 0.7)}}});
+  ncircuit.push_back({{normal, 1, 1.0, {Ry::Create(4, 1, 0.1)}}});
+  add_noise(5, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {IS::Create(6, 0, 1)}}});
+  add_noise(7, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {Ry::Create(8, 0, 0.4)}}});
+  ncircuit.push_back({{normal, 1, 1.0, {Rx::Create(8, 1, 0.7)}}});
+  add_noise(9, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {IS::Create(10, 0, 1)}}});
+  add_noise(11, p, ncircuit);
+  ncircuit.push_back({{KrausOperator<types::Gate>::kMeasurement, 1, 1.0,
+                       {gate::Measurement<types::Gate>::Create(12, {0, 1})}}});
+  add_noise(13, p, ncircuit);
+
+  return ncircuit;
+}
+
+void Run1(const types::NoisyCircuit& ncircuit,
+          const std::vector<double>& expected_results) {
+  unsigned num_qubits = 2;
+  unsigned num_reps = 20000;
+
+  auto measure = [](uint64_t r, const types::State& state,
+                    const std::vector<uint64_t>& stat,
+                    std::vector<unsigned>& histogram) {
+    ASSERT_EQ(stat.size(), 1);
+    ++histogram[stat[0]];
+  };
+
+  std::vector<unsigned> histogram(1 << num_qubits, 0);
+
+  types::QTSimulaltor::Parameter param;
+  param.collect_mea_stat = true;
+
+  EXPECT_TRUE(types::QTSimulaltor::Run(param, num_qubits, ncircuit,
+                                       0, num_reps, measure, histogram));
+
+  for (std::size_t i = 0; i < histogram.size(); ++i) {
+    EXPECT_NEAR(double(histogram[i]) / num_reps, expected_results[i], 0.005);
+  }
+}
+
+void Run2(const types::NoisyCircuit& ncircuit,
+          const std::vector<double>& expected_results) {
+  unsigned num_qubits = 2;
+  unsigned num_reps = 20000;
+
+  types::StateSpace state_space(1);
+
+  types::State scratch = state_space.Null();
+  types::State state = state_space.Create(num_qubits);
+  EXPECT_FALSE(state_space.IsNull(state));
+
+  std::vector<uint64_t> stat;
+
+  std::vector<unsigned> histogram(1 << num_qubits, 0);
+
+  types::QTSimulaltor::Parameter param;
+
+  for (unsigned i = 0; i < num_reps; ++i) {
+    state_space.SetStateZero(state);
+    param.collect_mea_stat = true;
+
+    EXPECT_TRUE(types::QTSimulaltor::Run(param, num_qubits, ncircuit, i,
+                                         scratch, state, stat));
+
+    ASSERT_EQ(stat.size(), 1);
+    ++histogram[stat[0]];
+  }
+
+  for (std::size_t i = 0; i < histogram.size(); ++i) {
+    EXPECT_NEAR(double(histogram[i]) / num_reps, expected_results[i], 0.005);
+  }
+}
+
+}  // namespace
+
+TEST(QTrajectoryTest, BitFlip) {
+/* The expected results are obtained with the following Cirq code.
```

Comment:
[No change required] Thanks for these code snippets - it will make it much easier to construct the Cirq-to-qsim translation layer :)

Reply:
Good.